RepositoryCompiler/CodeModel/CodeParsers/CSharp/CaDETMemberMetricCalculator.cs

```diff
@@ -121,5 +137,404 @@ private int CountOccurrences(string text, string pattern)
+        {
+            return method.DescendantNodes().OfType<TryStatementSyntax>().Count();
+        }
+
+        private int CountLoops(MemberDeclarationSyntax method)
+        {
+            int count = method.DescendantNodes().OfType<ForEachStatementSyntax>().Count();
+            count += method.DescendantNodes().OfType<ForStatementSyntax>().Count();
+            count += method.DescendantNodes().OfType<WhileStatementSyntax>().Count();
+            count += method.DescendantNodes().OfType<DoStatementSyntax>().Count();
+            return count;
+        }
+
+        private int CountReturnStatements(MemberDeclarationSyntax method)
+        {
+            return method.DescendantNodes().OfType<ReturnStatementSyntax>().Count();
+        }
+
+        private int CountComparisonOperators(MemberDeclarationSyntax method)
+        {
+            return method.DescendantNodes().OfType<BinaryExpressionSyntax>()
+                .Count(n => n.IsKind(SyntaxKind.EqualsExpression) || 
+                            n.IsKind(SyntaxKind.NotEqualsExpression) ||
+                            n.IsKind(SyntaxKind.LessThanExpression) ||
+                            n.IsKind(SyntaxKind.LessThanOrEqualExpression) ||
+                            n.IsKind(SyntaxKind.GreaterThanExpression) ||
+                            n.IsKind(SyntaxKind.GreaterThanOrEqualExpression));
+        }
+
+        private int CountMethodInvocations(MemberDeclarationSyntax method)
+        {
+            return method.DescendantNodes().OfType<InvocationExpressionSyntax>().Count();
+        }
+
+        private int CountUniqueMethodInvocations(MemberDeclarationSyntax method)
+        {
+            List<string> uniqueInvokedMethods = new List<string>();
+            var allInvokedMethods = method.DescendantNodes().OfType<InvocationExpressionSyntax>();
+            foreach (var invokedMethod in allInvokedMethods)
+            {
+                if (!uniqueInvokedMethods.Contains(invokedMethod.Expression.ToString()))
+                {
+                    uniqueInvokedMethods.Add(invokedMethod.Expression.ToString());
+                }
+            }
+            return uniqueInvokedMethods.Count();
+        }
+
+        private int CountNumberOfAssignments(MemberDeclarationSyntax method)
+        {
+            return method.DescendantNodes().OfType<AssignmentExpressionSyntax>().Count();
+        }
+
+        private int CountNumberOfNumericLiterals(MemberDeclarationSyntax method)
+        {
+            return method.DescendantNodes().OfType<LiteralExpressionSyntax>().Count(n => n.IsKind(SyntaxKind.NumericLiteralExpression));
+        }
+
+        private int CountNumberOfStringLiterals(MemberDeclarationSyntax method)
+        {
+            return method.DescendantNodes().OfType<LiteralExpressionSyntax>().Count(n => n.IsKind(SyntaxKind.StringLiteralExpression));
+        }
+
+        private int CountNumberOfMathOperations(MemberDeclarationSyntax method)
```

Comment:
This will also increment for "Pera " + "Peric". Is that what we want?

Reply:
I could change the code so that it does not include string concatenation. The question is whether concatenation would actually complicate the method/class? If there is a "reasonable" number of concatenations, then probably not, but I could imagine a method with a large number of concatenations that would require the programmer to stop and take the time to interpret the meaning of the code.